### PR TITLE
Fix is canvas kit bool

### LINKED ISSUE #116693: [Web][Framework] `isCanvasKit` implementation is not correct
the bool `isCanvasKit` is not correct for the case when the web renderer is auto. 

https://github.com/flutter/flutter/blob/6e5582da99f3c954c1f969c3ae4437bc0804400a/packages/flutter/lib/src/foundation/constants.dart#L74

### PR DESCRIPTION
**Pull request by alanwutang11**
fixes: https://github.com/flutter/flutter/issues/116693

uses integration test to verify implementation is correct 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/foundation/capabilities.dart [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import '_capabilities_io.dart'
+  if (dart.library.js_util) '_capabilities_web.dart' as capabilities;
+
+/// A bool that is True if the application is using canvasKit. Only to be used for web.
+bool isCanvasKit = capabilities.isCanvasKit;
```
Comment (kevmoo) [inline packages/flutter/lib/src/foundation/capabilities.dart:9]: Is this meant to be `set`? Should this be a getter?

Comment (alanwutang11) [reply packages/flutter/lib/src/foundation/capabilities.dart:9]: yup, it should be a getter. Thanks for pointing that out!

File: packages/flutter/lib/src/foundation/capabilities.dart [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import '_capabilities_io.dart'
+  if (dart.library.js_util) '_capabilities_web.dart' as capabilities;
+
+/// A bool that is True if the application is using canvasKit. Only to be used for web.
```
Comment (yjbanov) [inline packages/flutter/lib/src/foundation/capabilities.dart:8]: Make the first sentence a stand-alone paragraph.

See: https://dart.dev/guides/language/effective-dart/documentation#do-separate-the-first-sentence-of-a-doc-comment-into-its-own-paragraph

Comment (yjbanov) [inline packages/flutter/lib/src/foundation/capabilities.dart:8]: nit: s/A bool that is True/Returns true/
nit: s/canvasKit/CanvasKit/

File: packages/flutter/test/foundation/capabilities_test.dart [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+void main() {
+  test('isCanvasKit is true for CanvasKit', () {
+    expect(isCanvasKit, false);
+  }, skip: isCanvasKit); // [intended] isCanvasKit is what we are testing here.
```
Comment (yjbanov) [inline packages/flutter/test/foundation/capabilities_test.dart:11]: This test has no effect. Let's say `isCanvasKit` returns a wrong value. For example, we are in CanvasKit mode but the bool returns false. Then the test will run and `expect(isCanvasKit, false)` will pass. The opposite is the same. Therefore the test does not actually test anything.

File: dev/integration_tests/web_e2e_tests/test_driver/capabilities_integration.dart [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+void main() {
+  IntegrationTestWidgetsFlutterBinding.ensureInitialized();
+
+  // App is run using 'web-renderer=auto'. Therefore, when the targetPlatform isDesktop,
+  // the app should be using canvasKit.
+  final bool isDesktop = _isDesktop(defaultTargetPlatform);
+
+  testWidgets('Capabilities integration test',
+          (WidgetTester tester) async {
+
```
Comment (yjbanov) [inline dev/integration_tests/web_e2e_tests/test_driver/capabilities_integration.dart:20]: nit: unnecessary empty line

File: dev/integration_tests/web_e2e_tests/test_driver/capabilities_integration.dart [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+void main() {
+  IntegrationTestWidgetsFlutterBinding.ensureInitialized();
+
+  // App is run using 'web-renderer=auto'. Therefore, when the targetPlatform isDesktop,
+  // the app should be using canvasKit.
+  final bool isDesktop = _isDesktop(defaultTargetPlatform);
+
+  testWidgets('Capabilities integration test',
+          (WidgetTester tester) async {
```
Comment (yjbanov) [inline dev/integration_tests/web_e2e_tests/test_driver/capabilities_integration.dart:19]: nit: this has too much indent. I'd actually leave it on the previous line.

File: dev/bots/test.dart [hunk 6]
```diff
@@ -1177,6 +1177,9 @@ Future<void> _runWebLongRunningTests() async {
     () => _runWebE2eTest('url_strategy_integration', buildMode: 'profile', renderer: 'canvaskit'),
     () => _runWebE2eTest('url_strategy_integration', buildMode: 'release', renderer: 'html'),
 
+    // This test only needs to test when renderer=auto
+    () => _runWebE2eTest('capabilities_integration', buildMode: 'debug', renderer: 'auto'),
```
Comment (yjbanov) [inline dev/bots/test.dart:1181]: Is this sufficient? I thought you also wanted to test other modes.

File: packages/flutter/test/foundation/capabilities_test.dart (hunk 3 above)
Comment (alanwutang11) [reply packages/flutter/test/foundation/capabilities_test.dart:11]: ah, that makes sense. I copied the `kIsWeb` test without much thought...

File: dev/bots/test.dart (hunk 6 above)
Comment (alanwutang11) [reply dev/bots/test.dart]: Added

File: packages/flutter/test/foundation/capabilities_test.dart (hunk 3 above)
Comment (yjbanov) [reply packages/flutter/test/foundation/capabilities_test.dart]: Haha, maybe `kIsWeb` isn't testing anything either :) Perhaps what this test could test is that `isCanvasKit` doesn't crash. Correctness can be covered by the integration test.

File: packages/flutter/lib/src/foundation/_capabilities_io.dart [hunk 7]
```diff
@@ -0,0 +1,8 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/// The dart:io implementation of [isCanvasKit]. This bool shouldn't be used outside of web.
```
Comment (yjbanov) [inline packages/flutter/lib/src/foundation/_capabilities_io.dart:5]: Make the first sentence a standalone paragraph.

File: dev/integration_tests/web_e2e_tests/test_driver/capabilities_integration_canvaskit.dart [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:integration_test/integration_test.dart';
+
+void main() {
+  IntegrationTestWidgetsFlutterBinding.ensureInitialized();
+
+  testWidgets('Capabilities integration test', (WidgetTester tester) async {
```
Comment (yjbanov) [inline dev/integration_tests/web_e2e_tests/test_driver/capabilities_integration_canvaskit.dart:12]: More precise test descriptions are more useful. Suggestion: "isCanvasKit returns true in CanvasKit mode"

File: dev/integration_tests/web_e2e_tests/test_driver/capabilities_integration_html.dart [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:integration_test/integration_test.dart';
+void main() {
+  IntegrationTestWidgetsFlutterBinding.ensureInitialized();
+
+  testWidgets('Capabilities integration test', (WidgetTester tester) async {
```
Comment (yjbanov) [inline dev/integration_tests/web_e2e_tests/test_driver/capabilities_integration_html.dart:11]: Similarly here "isCanvasKit returns false in HTML mode"

File: packages/flutter/lib/src/foundation/_capabilities_web.dart [hunk 10]
```diff
@@ -0,0 +1,13 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:js/js.dart';
+
+/// This value is set by the engine. It is used to determine if the application is
+/// using canvaskit.
+@JS('window.flutterCanvasKit')
+external Object? get windowFlutterCanvasKit;
```
Comment (yjbanov) [inline packages/flutter/lib/src/foundation/_capabilities_web.dart:10]: The can be private, I think.

**Review by yjbanov — APPROVED** [review]
LGTM with a few nitpicks.

File: packages/flutter/lib/src/foundation/_capabilities_web.dart (hunk 10 above)
Comment (alanwutang11) [reply packages/flutter/lib/src/foundation/_capabilities_web.dart:10]: agree

**Pull request merged**